### ISSUE TITLE
firecracker: pass machine configuration

### ISSUE THREAD

**Issue by niektemme** [kata-containers/runtime#1594]
Is there a way to pass machine-config to the firecracker runtime? Specifically, memory and vcpu count. I looked at fc.go and as far as I can tell these parameters are not pushed to the firecracker golang sdk. 

So far, my experience with kata and the firecracker runtime have been great! Reducing the memory footprint of my test container form 170MB using qemu-light to 110MB using firecracker. 

My goal is to reduce the memory footprint of each container (increase density). So far, I have not been able to run a container with less than 110MB of memory footprint. The same container takes 5mb memory when using runc. My assumption is that reducing the machine memory size of the microvm from the default 128MB to 64MB or lower, would also reduce the memory footprint of the container.

This would be similar to applying machine config directly in firecracker using
```
curl --unix-socket /tmp/firecracker.socket -i  \
    -X PUT 'http://localhost/machine-config' \
    -H 'Accept: application/json'            \
    -H 'Content-Type: application/json'      \
    -d '{
        "vcpu_count": 1,
        "mem_size_mib": 64
    }'
```
As described https://github.com/firecracker-microvm/firecracker/blob/master/docs/getting-started.md
Or using the -m option in the firecracker firectl.

I was able to successfully install and run kata containers with firecracker support (kata 1.5). I followed the instructions on the documentation page. https://github.com/kata-containers/documentation/wiki/Initial-release-of-Kata-Containers-with-Firecracker-support

Except for adding the runtime. Instead of creating the docker daemon.json configuration file, as described in the documentation page, I added the runtime in the systemd docker.service config file. I originally followed the instructions of automated install of kata, which already added the kata-runtime in the systemd docker config.

```
[Service]
ExecStart=
ExecStart=/usr/bin/dockerd -D --add-runtime kata-fc=/opt/kata/bin/kata-fc --add-runtime kata-runtime=/usr/bin/kata-runtime --default-runtime=kata-runtime --storage-driver=devicemapper
```

Running containers with the kata-fc runtime works as described in the kata firecracker documentation page. As a test case I create a simple go program. A basic web server returning a hello world page. The go program is compiled in a Dockerfile and added to a standard alpine:latest image. 

Each container created using this image reduces the available memory by about 110MB. As can be seen in the output below (part), which shows the available memory (using free -m) after creating each container. This output is created using a shell script that automatically creates one container and then shows the free -m output after each container is created.

```
….
Welcome 110 times
26e30739d09347fe7f4853daa4d38a02717a91e3ddac9bdf35c8bbb8515ec6b9
              total        used        free      shared  buff/cache   available
Mem:          15993        2195        1693        9087       12104        4383
Swap:          4095           0        4095
Welcome 111 times
c2b937ada952542a408ad403cda2707f3dc705cea37aa2fed0f5d99420c36ac5
              total        used        free      shared  buff/cache   available
Mem:          15993        2212        1582        9169       12198        4284
Swap:          4095           0        4095
Welcome 112 times
d339b2f1a1aea10f2a676beef8582a996d59c052a3a53746120aee89cb6ffd04
              total        used        free      shared  buff/cache   available
Mem:          15993        2230        1469        9250       12292        4184
Swap:          4095           0        4095
….

```
My assumption is that reducing the firecracker mem_size_mib to 64MB would also reduce the memory footprint of each container. Of course, I could be wrong in this assumption and perhaps do not correctly understand how kata interacts with firecracker.

I did check the firecracker instances created by kata to see their settings. Simply by using curl to access them. This shows that the mem_size_mib is 128MB. This could also just showing be the default, as I of course am not able to put a different size to a running microvm. See output bellow. 

```
sudo curl --unix-socket /run/vc/sbs/07a4fae3bfd867ac0e245aee57c524c6b4a6b331ec22b58b2fa4cd476a9f3f3d/firecracker.sock -i  \
>     -X GET 'http://localhost/machine-config' \
>     -H 'Accept: application/json'            \
>     -H 'Content-Type: application/json'
HTTP/1.1 200 OK
Content-Type: application/json
Transfer-Encoding: chunked
Date: Thu, 25 Apr 2019 07:12:53 GMT

{ "vcpu_count": 1, "mem_size_mib": 128,  "ht_enabled": false,  "cpu_template": "Uninitialized" } 
```

I also looked at the fc.go file myself, which I assume contains the code for creating the firecracker instance. Unfortunately, I am not an experienced go programmer.  

My assumption is that it would require an additional function. Something like
```
func (fc *firecracker) fcSetMachineConfiguration(cpuTemplate CPUTemplate, htEnabled bool, memSizeMib int64, vcpuCount int64) error {
	span, _ := fc.trace("fcSetMachineConfiguration")
	defer span.Finish()
	fc.Logger().WithField("Machine-configuration-htEnabled", htEnabled).Debug()
	fc.Logger().WithField("Machine-configuration-memSizeMib", memSizeMib).Debug()
	fc.Logger().WithField("Machine-configuration-vcpuCount", vcpuCount).Debug()


	machineParams := ops.NewPutMachineConfigurationParams()

 machine := &models.MachineConfiguration{
	  CPUTemplate: &cpuTemplate,
		HtEnabled: &tEnabled,
		MemSizeMib: &memSizeMib,
		VcpuCount: &vcpuCount,
	}

machineParams.SetBody(machine)
	_, err := fc.client().Operations.PutMachineConfiguration(machineParams)
	if err != nil {
		return err
	}

	return nil
}
```

**Comment by grahamwhaley** [comment]
Hi @niektemme - the (default) memory passed to a container (VM) in Kata is normally defined in the kata config toml file. if you `kata-fc kata-env`, that should tell you the path to the config file, and you can see the template with the memory size line at https://github.com/kata-containers/runtime/blob/master/cli/config/configuration-fc.toml.in#L67

I don't know many actual fc details wrt the kata integration though, so /cc @egernst @sboeuf for any more thoughts.

For the memory reduction test, we do have a test file that will do that for you ;-) - see https://github.com/kata-containers/tests/tree/master/metrics/density and https://github.com/kata-containers/tests/tree/master/metrics/report if you want some pretty graphs of the results. The `footprint_data` test captures 'free' and `/proc/meminfo` data into the json results file - and the report generator will then extract and plot some of that. It will probably show the same result you already have though ;-) OOI, are you looking at 'free' or 'available' - the difference being iirc one will take into account data that is in caches that could be freed up if needed. As kata has a bigger transient footprint, it might be that it has filled the caches a bit more, but that space could be available if needed iyswim.

Oh, and, thanks for the detailed query! ;-)

**Comment by devimc** [comment]
> So far, my experience with kata and the firecracker runtime have been great! Reducing the memory footprint of my test container form 170MB using qemu-light to 110MB using firecracker.

sweet, but I think FC is not fully compatible with kubernetes. You can have a good memory footprint with Qemu4 [1][2] and kata 1.6.x.  Set `default_memory = 512` and `kernel_params = "init=/usr/bin/kata-agent"` in the configuration file to have better numbers.

[1] - https://github.com/kata-containers/qemu/tree/kata-qemu-4.0.0
[2] - https://github.com/kata-containers/packaging/blob/master/scripts/configure-hypervisor.sh

**Comment by sboeuf** [comment]
@mcastelino I was looking into @niektemme issue here and I went to look at the `fc.go` code, and that's where I realized we don't implement for Firecracker the way to setup a specific amount of CPU with a specific amount of memory. Am I missing something?
If that's the case, then `fc.go` needs to be extended to make sure it can be started with other parameters than the default ones picked by Firecracker (1 vCPU and 128MiB of RAM).
